app/services/ems_dashboard_service.rb

```diff
@@ -33,15 +33,36 @@ def recent_resources(model, title, label)
     }.compact
   end
 
+  # Generates the following SQL:
+  #
+  #   SELECT COUNT(id), to_char(date_trunc('day', created_on), 'YYYY-MM-DD')
+  #   FROM   model
+  #   WHERE  ems_id = ?
+  #     AND (created_on > 30.days.ago)
+  #   GROUP BY to_char(date_trunc('day', created_on), 'YYYY-MM-DD')
+  #
+  # Returns in the format:
+  #
+  #   {
+  #     "2000-01-01" => 123,
+  #     "2000-01-02" => 234,
+  #     "2000-01-03" => 345,
+  #     ...
+  #   }
   def recent_records(model)
-    all_records = Hash.new(0)
-    records = model.where('created_on > ? and ems_id = ?', 30.days.ago.utc, @ems.id)
-    records = records.includes(:resource => [:ext_management_system]) if @ems.blank?
-    records.sort_by(&:created_on).uniq.each do |r|
-      date = r.created_on.strftime("%Y-%m-%d")
-      all_records[date] += model.where('created_on = ?', r.created_on).count
-    end
-    all_records
+    db_table     = model.arel_table
+    trunc_args   = [Arel::Nodes::SqlLiteral.new("'day'"), db_table[:created_on]]
+    date_trunc   = Arel::Nodes::NamedFunction.new("date_trunc", trunc_args)
+    to_char_args = [date_trunc, Arel::Nodes::SqlLiteral.new("'YYYY-MM-DD'")]
+    group_by_sql = Arel::Nodes::NamedFunction.new("to_char", to_char_args)
```

Comment:
So... about this...

To start off, this is what it used to be:

```ruby
CREATED_ON_SELECT = %{to_char("%{table}"."created_on", 'YYYY-MM-DD') as created_on_fmt}.freeze
def recent_records(model)
  sql = model.select(CREATED_ON_SELECT % {:table => model.table_name})
             .select("count(id)")
             .where(:ems_id => @ems.id)
             .where("created_on > ?", 30.days.ago.utc)
             .group("created_on_fmt")
  sql = sql.includes(:resource => [:ext_management_system]) if @ems.blank?

  ActiveRecord::Base.connection.select_all(sql.to_sql).rows.to_h
end
```

But, that `CREATED_ON_SELECT % {:table => model.table_name}` bit is technically SQL injection, and I didn't like `sprintf` stuff I did anyway, so I changed it to the following:

```diff
-  CREATED_ON_SELECT = %{to_char("%{table}"."created_on", 'YYYY-MM-DD') as created_on_fmt}.freeze
   def recent_records(model)
-    sql = model.select(CREATED_ON_SELECT % {:table => model.table_name})
-               .select("count(id)")
+    db_table     = model
+    to_char_args = [db_table[:created_on], Arel::Nodes::SqlLiteral.new("'YYYY-MM-DD'")]
+    arel_select  = Arel::Nodes::NamedFunction.new("to_char", to_char_args)
+
+    sql = model.select(select.as("created_on_fmt")
+               .select(db_table[:id].count)
                .where(:ems_id => @ems.id)
-               .where("created_on > ?", 30.days.ago.utc)
+               .where(db_table[:created_on].gt(30.days.ago.utc))
                .group("created_on_fmt")
     sql = sql.includes(:resource => [:ext_management_system]) if @ems.blank?
+
     ActiveRecord::Base.connection.select_all(sql.to_sql).rows.to_h
   end
```

And then I realized we had talked about changing this whole thing to using a [`.count` based approach](https://github.com/ManageIQ/manageiq-ui-classic/pull/4447#issuecomment-412102442), so what you are seeing here above is basically:

```ruby
.group("date_trunc('day', created_on)")
```

But done in a way that we are always doing a `"vms"."created_on"` (or whatever model) for all the places where it is done (avoids issues if this ever gets turned into a function and there are shared columns with joins), and it avoids needing to anything to avoid SQL inject.  It becomes less of a "one-liner", but that is why it is in a function I guess....

Reply:
I tend to not like going the arel route, this is much easier to read.

(Yes, you go on to say you are avoiding this route - I'm just saying if you end up backtracking, I like this one too)

Reply:
surprised you need the `to_char`, thought the `date_trunc` would be good enough - but it probably is safest to keep the code as similar as possible.

Reply:
> I tend to not like going the arel route, this is much easier to read.

I assume you meant "but this is much easier to read", which I guess is fine.  I was mostly trying to avoid SQL injection warnings, but if you think it is easier to grok, I guess that is cool.

> surprised you need the `to_char`

@kbrock I guess I didn't mention this above, but this is to make sure we are keeping consistency with what is expect for the output of this method, which is (as the comment in the code shows above):

```ruby
{
  "2000-01-01" => 123,
  "2000-01-02" => 234,
  "2000-01-03" => 345,
  # ...
}
```

You can see what happens when I don't have the `to_char` with this sample console statement:

```ruby
irb(main):001:0> db_table = MiqTemplate.arel_table; nil
#=> nil
irb(main):002:0> trunc_args   = [Arel::Nodes::SqlLiteral.new("'day'"), db_table[:created_on]]; nil
#=> nil
irb(main):003:0> date_trunc   = Arel::Nodes::NamedFunction.new("date_trunc", trunc_args); nil
#=> nil
irb(main):004:0> to_char_args = [date_trunc, Arel::Nodes::SqlLiteral.new("'YYYY-MM-DD'")]; nil
#=> nil
irb(main):005:0> group_by_sql = Arel::Nodes::NamedFunction.new("to_char", to_char_args); nil
#=> nil
irb(main):006:0> MiqTemplate.where(:ems_id => 1).where(db_table[:created_on].gt(30.days.ago.utc)).group(group_by_sql.to_sql).count
#   (159.2ms)  SELECT COUNT(*) AS count_all,
#                      to_char(date_trunc('day', "vms"."created_on"), 'YYYY-MM-DD') AS to_char_date_trunc_day_vms_created_on_yyyy_mm_dd
#              FROM "vms"
#              WHERE "vms"."type" IN ('MiqTemplate', 'ManageIQ::Providers::CloudManager::Template', 'ManageIQ::Providers::InfraManager::Template', 'ManageIQ::Providers::Amazon::CloudManager::Template', 'ManageIQ::Providers::Azure::CloudManager::Template', 'ManageIQ::Providers::Google::CloudManager::Template', 'ManageIQ::Providers::Openstack::CloudManager::Template', 'ManageIQ::Providers::Openstack::CloudManager::VolumeSnapshotTemplate', 'ManageIQ::Providers::Openstack::CloudManager::VolumeTemplate', 'ManageIQ::Providers::Vmware::CloudManager::Template', 'TemplateXen', 'ManageIQ::Providers::Kubevirt::InfraManager::Template', 'ManageIQ::Providers::Openstack::InfraManager::Template', 'ManageIQ::Providers::Redhat::InfraManager::Template', 'ManageIQ::Providers::Microsoft::InfraManager::Template', 'ManageIQ::Providers::Vmware::InfraManager::Template')
#                AND "vms"."template" = $1
#                AND "vms"."ems_id" = $2
#                AND ("vms"."created_on" > '2018-07-24 16:38:53.075362')
#              GROUP BY to_char(date_trunc('day', "vms"."created_on"), 'YYYY-MM-DD')  [["template", "t"], ["ems_id", 1]]
#=> {"2018-08-09"=>76864}
irb(main):007:0> MiqTemplate.where(:ems_id => 1).where(db_table[:created_on].gt(30.days.ago.utc)).group(date_trunc.to_sql).count
#   (84.1ms)  SELECT COUNT(*) AS count_all,
#                    date_trunc('day', "vms"."created_on") AS date_trunc_day_vms_created_on
#             FROM "vms"
#             WHERE "vms"."type" IN ('MiqTemplate', 'ManageIQ::Providers::CloudManager::Template', 'ManageIQ::Providers::InfraManager::Template', 'ManageIQ::Providers::Amazon::CloudManager::Template', 'ManageIQ::Providers::Azure::CloudManager::Template', 'ManageIQ::Providers::Google::CloudManager::Template', 'ManageIQ::Providers::Openstack::CloudManager::Template', 'ManageIQ::Providers::Openstack::CloudManager::VolumeSnapshotTemplate', 'ManageIQ::Providers::Openstack::CloudManager::VolumeTemplate', 'ManageIQ::Providers::Vmware::CloudManager::Template', 'TemplateXen', 'ManageIQ::Providers::Kubevirt::InfraManager::Template', 'ManageIQ::Providers::Openstack::InfraManager::Template', 'ManageIQ::Providers::Redhat::InfraManager::Template', 'ManageIQ::Providers::Microsoft::InfraManager::Template', 'ManageIQ::Providers::Vmware::InfraManager::Template')
#               AND "vms"."template" = $1
#               AND "vms"."ems_id" = $2
#               AND ("vms"."created_on" > '2018-07-24 16:39:06.163190')
#             GROUP BY date_trunc('day', "vms"."created_on")  [["template", "t"], ["ems_id", 1]]
#=> {2018-08-09 00:00:00 UTC=>76864}
```

So if without the `to_char`, we get a key of `Date`, not of `String`.  If this was a scope, I might have considered keeping it as a date, but because the hash is just meant to be a data value for a chart, I figured it made more sense to just have SQL format it for us (since it would be quicker at it).

It slows down the query performance a bit, but I assume that is quicker than us trying to format all of the keys in memory via ruby.